src/nnp_environment_types.F

```diff
@@ -0,0 +1,618 @@
+!> \param molecule_kind_set The set of all different molecule kinds involved
+!> \param molecule_set The set of all molecules
+!> \param local_molecules All molecules on this particular node
+!> \param nnp_input ...
+!> \param force_env_input Pointer to the force_env input section
+!> \param cell_ref The reference simulation cell
+!> \param use_ref_cell Logical which indicates if reference
+!>                      simulation cell is used
+!> \param nnp_potential_energy ...
+!> \date   2020-10-10
+!> \author Christoph Schran (christoph.schran@rub.de)
+!> \note
+!>   For possible missing arguments see the attributes of nnp_type
+! **************************************************************************************************
+   SUBROUTINE nnp_env_set(nnp_env, nnp_forces, subsys, &
+                          atomic_kind_set, particle_set, local_particles, &
+                          molecule_kind_set, molecule_set, local_molecules, &
+                          nnp_input, force_env_input, cell_ref, &
+                          use_ref_cell, nnp_potential_energy)
+
+      TYPE(nnp_type), POINTER                            :: nnp_env
+      REAL(KIND=dp), DIMENSION(:, :), OPTIONAL, POINTER  :: nnp_forces
+      TYPE(cp_subsys_type), OPTIONAL, POINTER            :: subsys
+      TYPE(atomic_kind_type), DIMENSION(:), OPTIONAL, &
+         POINTER                                         :: atomic_kind_set
+      TYPE(particle_type), DIMENSION(:), OPTIONAL, &
+         POINTER                                         :: particle_set
+      TYPE(distribution_1d_type), OPTIONAL, POINTER      :: local_particles
+      TYPE(molecule_kind_type), DIMENSION(:), OPTIONAL, &
+         POINTER                                         :: molecule_kind_set
+      TYPE(molecule_type), DIMENSION(:), OPTIONAL, &
+         POINTER                                         :: molecule_set
+      TYPE(distribution_1d_type), OPTIONAL, POINTER      :: local_molecules
+      TYPE(section_vals_type), OPTIONAL, POINTER         :: nnp_input, force_env_input
+      TYPE(cell_type), OPTIONAL, POINTER                 :: cell_ref
+      LOGICAL, INTENT(IN), OPTIONAL                      :: use_ref_cell
+      REAL(KIND=dp), INTENT(IN), OPTIONAL                :: nnp_potential_energy
+
+      CHARACTER(len=*), PARAMETER :: routineN = 'nnp_env_set', routineP = moduleN//':'//routineN
+
+      TYPE(atomic_kind_list_type), POINTER               :: atomic_kinds
+      TYPE(molecule_kind_list_type), POINTER             :: molecule_kinds
+      TYPE(molecule_list_type), POINTER                  :: molecules
+      TYPE(particle_list_type), POINTER                  :: particles
+
+      CPASSERT(ASSOCIATED(nnp_env))
+      CPASSERT(nnp_env%ref_count > 0)
+
+      IF (PRESENT(nnp_potential_energy)) THEN
+         nnp_env%nnp_potential_energy = nnp_potential_energy
+      END IF
+      IF (PRESENT(nnp_forces)) nnp_env%nnp_forces = nnp_forces
```

Comment:
When assigning to an ALLOCATABLE please use `(:)` to prevent automatic reallocations, like this:

```
 IF (PRESENT(nnp_forces)) nnp_env%nnp_forces(:) = nnp_forces
```

Reply:
Ohh, ok! Got it, thanks!